app/src/androidTest/java/com/sdp13epfl2021/projmag/database/FakeDatabaseTest.kt

```diff
@@ -0,0 +1,102 @@
+package com.sdp13epfl2021.projmag.database
+
+import com.sdp13epfl2021.projmag.model.ImmutableProject
+import java.util.*
+
+
+class FakeDatabaseTest(projectsBeginning: List<ImmutableProject>) : ProjectsDatabase {
+
+    var projects: List<ImmutableProject> = projectsBeginning
+    var listeners: List<((ProjectChange) -> Unit)> = emptyList()
+    var nextId: Int = 90000
+
+    fun add(project: ImmutableProject) {
+        projects = projects + project
+        listeners.forEach{ it -> it(ProjectChange(ProjectChange.Type.ADDED, project)) }
+    }
+
+    fun modify(project: ImmutableProject) {
+        projects.filter { p -> p.id != project.id }
+        projects = projects + project
+        listeners.forEach{ it -> it(ProjectChange(ProjectChange.Type.MODIFIED, project)) }

```

Comment:
1)Why are you not reaffecting the value to project in projects.filter
2) Why are you not checking that filter did find a previously existing project before adding the modified project (meaning if this was a project that did not exist in the projects list, then modify will do exactly the same as the add function, is this wanted behaviour?)

Reply:
1) It was a mistake. Thanks for reporting that error.
2) The modify method (inside the test class) is used to trigger a MODIFIED change, and yes the effect on projects is the same as the add.

Reply:
But if we have :

Database(project(A), project(B))

and we call modify on this database with a project(C)
the database becomes Database(A,B,C), adding a project, but the listeners receive a MODIFIED notice instead of an ADD notice ? no?

Reply:
Yes that's correct. And this can be the case in the real ProjectDatabase, for example : 
We have p1, p2 in our phone, we lost connection, p3 is added to firebase, we reconnect, p3 is modified => we will receive modified.